### PR TITLE
[wip] Save `estimated_norm_scaling_factor` in checkpoints

### PR DESCRIPTION
**Pull request by oli-clive-griffin**
description coming soon

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [ ] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

### Performance Check.

If you have implemented a training change, please indicate precisely how performance changes with respect to the following metrics:
- [ ] L0
- [ ] CE Loss
- [ ] MSE Loss
- [ ] Feature Dashboard Interpretability

Please links to wandb dashboards with a control and test group.

### CODE REVIEW

File: sae_lens/sae_training_runner.py [hunk 1]
```diff
@@ -187,58 +183,57 @@ def _init_sae_group_b_decs(
...
-        config = trainer.cfg.to_dict()
-        with open(f"{path}/cfg.json", "w") as f:
+        config = self.cfg.to_dict()
+        with open(cfg_path, "w") as f:
             json.dump(config, f)
 
-        log_feature_sparsities = {"sparsity": trainer.log_feature_sparsity}
-
-        log_feature_sparsity_path = f"{path}/{SPARSITY_PATH}"
-        save_file(log_feature_sparsities, log_feature_sparsity_path)
```
Comment (oli-clive-griffin) [inline sae_lens/sae_training_runner.py:214]: `self.sae.save_model` already does this if you pass the sparsity tensor in. we can remove duplicate/coupled code this way

File: sae_lens/sae_training_runner.py [hunk 2]
```diff
@@ -187,58 +183,57 @@ def _init_sae_group_b_decs(
...
+        self.activations_store.save_state(base_path)
 
+        # TODO(oli-clive-griffin): Is this broken? it seems like this is "assymetrical" in the sense that
+        # this changes the output of the SAE.
+        # Should we not also balance this by scaling the encoder weights?
         if self.sae.cfg.normalize_sae_decoder:
             self.sae.set_decoder_norm_to_unit_norm()
-        self.sae.save_model(path)
+
+        weights_path, cfg_path, sparsity_path = self.sae.save_model(
```
Comment (oli-clive-griffin) [inline sae_lens/sae_training_runner.py:202]: by returning these paths, we're able to remove this module's knowledge of the internal mechanism of `self.sae.save_model` when reading the files it saves below. i.e. the below code doesn't need to know that the correct path is: `f"{path}/{SAE_WEIGHTS_PATH}"` in `model_artifact.add_file(f"{path}/{SAE_WEIGHTS_PATH}")`.

File: sae_lens/training/activations_store.py [hunk 3]
```diff
@@ -700,16 +712,13 @@ def next_batch(self):
...
+    def save_state(self, path: Path):
+        state = {
+            "norm_scaling_factor": self.estimated_norm_scaling_factor,
         }
-        if self._storage_buffer is not None:  # first time might be None
-            result["storage_buffer"] = self._storage_buffer
-        return result
 
-    def save(self, file_path: str):
-        save_file(self.state_dict(), file_path)
```
Comment (oli-clive-griffin) [inline sae_lens/training/activations_store.py:712]: Don't read this as a rewrite of the removed functions, it's an unrelated removal and addition. The removed ones were unused as far as I can tell, but if they're a desired public interface I'll put them back in!

File: sae_lens/training/sae_trainer.py [hunk 4]
```diff
@@ -204,23 +212,14 @@ def fit(self) -> TrainingSAE:
...
 
-    @torch.no_grad()
-    def _estimate_norm_scaling_factor_if_needed(self) -> None:
-        if self.cfg.normalize_activations == "expected_average_only_in":
-            self.activation_store.estimated_norm_scaling_factor = (
-                self.activation_store.estimate_norm_scaling_factor()
-            )
-        else:
-            self.activation_store.estimated_norm_scaling_factor = 1.0
-
```
Comment (oli-clive-griffin) [inline sae_lens/training/sae_trainer.py:215]: I've moved this into `ActivationsStore` as it seemed more natural.

the general pattern of the change being:

```python
class ParentThing:
    def set_child_property():
        self.thing.property = self.thing.calculate_property()
```
vs
```python
class ChildThing
    def set_property():
        self.property = self.calculate_property()
```

File: sae_lens/training/sae_trainer.py [hunk 5]
```diff
@@ -204,23 +212,14 @@ def fit(self) -> TrainingSAE:
...
     def _train_step(
         self,
         sae: TrainingSAE,
         sae_in: torch.Tensor,
     ) -> TrainStepOutput:
 
         sae.train()
-        # Make sure the W_dec is still zero-norm
```
Comment (oli-clive-griffin) [inline sae_lens/training/sae_trainer.py:223]: I _think_ this was a typo. let me know if I've misunderstood

File: sae_lens/config.py [hunk 6]
```diff
@@ -29,6 +29,11 @@
 
 HfDataset = DatasetDict | Dataset | IterableDatasetDict | IterableDataset
 
+ActivationNormalizationStrategy = Literal[
+    "none", "expected_average_only_in", "constant_norm_rescale", "layer_norm"
+]
```
Comment (oli-clive-griffin) [inline sae_lens/config.py:34]: there were a few comments where the layer norm option was not noted. Though that centralising the definition would help

File: sae_lens/training/sae_trainer.py [hunk 7]
```diff
@@ -204,23 +212,14 @@ def fit(self) -> TrainingSAE:
...
+        # Make sure the W_dec is still unit-norm
```
Comment (oli-clive-griffin) [inline sae_lens/training/sae_trainer.py:222]: I _think_ this was a typo. let me know if I've misunderstood

**Comment by oli-clive-griffin** [comment]
closing this pr and resubmitting a more tightly scoped one as it's my first PR here
